### PR TITLE
Allow saving host if template URL returns 404

### LINKED ISSUE #47: Unable to save host if template URL returns not found
A user is unable to save a host if the template url is incorrect, e.g. the specified branch does not exist. Therefore a user cannot change the URL anymore and is in a dead-lock situation.

The error happens because before the host is saved `valid?` is called on the host object. This calls `validate_tftp` and this calls `host.provisioning_template` which is overridden in this plugin and tries to retrieve the template from git.

Possible solutions I can think of:
- validate the URL works before saving the host
- allow saving the host even if the template can not be downloaded

### PR DESCRIPTION
**Pull request by kamils-iRonin**
fix #47

### CODE REVIEW

**Comment by kamils-iRonin** [comment]
@timogoebel please check if this fixes the problem

File: app/models/concerns/foreman_git_templates/host_extensions.rb [hunk 1]
```diff
@@ -18,6 +18,9 @@ def git_template_tmpfile
       return unless git_template_url
 
       @git_template_tmpfile ||= RepositoryFetcher.call(git_template_url)
+    rescue ::ForemanGitTemplates::RepositoryFetcher::RepositoryFetcherError => e
+      logger.error "GitTemplates: #{e.message}"
```
Comment (timogoebel) [inline app/models/concerns/foreman_git_templates/host_extensions.rb]:
```suggestion
      Foreman::Logging.exception(“GitTemplates: Failed to fetch data from #{git_template_url}, e)
```

File: app/models/concerns/foreman_git_templates/host_extensions.rb [hunk 2]
```diff
@@ -18,6 +18,9 @@ def git_template_tmpfile
       return unless git_template_url
 
       @git_template_tmpfile ||= RepositoryFetcher.call(git_template_url)
+    rescue ::ForemanGitTemplates::RepositoryFetcher::RepositoryFetcherError => e
+      Foreman::Logging.exception(“GitTemplates: Failed to fetch data from #{git_template_url}, e)
```
Comment (timogoebel) [inline app/models/concerns/foreman_git_templates/host_extensions.rb]:
```suggestion
      Foreman::Logging.exception("GitTemplates: Failed to fetch data from #{git_template_url}", e)
```

^^ happens when you try to code on an iPad ;-)

File: test/models/foreman_git_templates/host_test.rb [hunk 3]
```diff
@@ -25,5 +25,78 @@ class HostTest < ActiveSupport::TestCase
...
+          end
+        end
+      end
+
+      context 'with invalid template_url' do
+        setup do
+          stub_request(:get, host.params['template_url']).to_return(status: 404)
+        end
+
+        let(:expected_errors) { ["No PXEGrub2 template was found for host #{host.name}. Repository url: #{host.params['template_url']}"] }
```
Comment (kamils-iRonin) [inline test/models/foreman_git_templates/host_test.rb]: @timogoebel I think this error is expected if the `template_url` is wrong. It's added here: https://github.com/dm-drogeriemarkt/foreman_git_templates/pull/48/files#diff-0a9b35b0c2c90515a22df356a448a6a1R30. So, it should not be possible to update the host with wrong `template_url`.

**Comment by kamils-iRonin** [comment]
@dgoetz @timogoebel I've added `template_url` validaton with HEAD request. Tested with GitLab and GitHub links and seems to work well. Please take a look.

**Comment by dgoetz** [comment]
I talked to @timogoebel in person and he thinks a user should be forced to change an invalid url, so a check is also required if the template_url is not changed.

**Comment by kamils-iRonin** [comment]
@dgoetz I'm not sure how we can force a user to update an invalid url. @timogoebel do you have any idea?

**Comment by timogoebel** [comment]
@kamils-iRonin: The idea was to invalidate the object. I think the code should already do that, I'm wondering why it does not work when the URL is wrong in the first place. Looks like the validation on the parameter is not run when the parameter value is not changed.

**Comment by kamils-iRonin** [comment]
@dgoetz Can you please test it again? It should not be possible to update a host without fixing invalid url now.

**Comment by timogoebel** [comment]
Thanks, @kamils-iRonin and @dgoetz.

**Pull request merged**